Python/bltinmodule.c

```diff
@@ -2971,7 +2971,12 @@ zip_next(zipobject *lz)
 
     if (tuplesize == 0)
         return NULL;
-    if (Py_REFCNT(result) == 1) {
+#ifdef Py_GIL_DISABLED
+    int reuse_tuple = 0;
+ #else
+    int reuse_tuple = Py_REFCNT(result) == 1;
+#endif
+    if (reuse_tuple) {
```

Comment:
I think we want to enable re-use under certain conditions. We should probably put the logic in some internal function, possibly in `pycore_object.h`. I think the condition for the free-threaded build is:

1) ob_tid matches `_Py_ThreadId()`
2) ob_ref_local is 1
3) ob_ref_shared is 0

The logic is that no other thread calling `zip_next` can re-use `lz->result` because of the thread id condition (condition 1). And no thread outside of `zip_next` can incref that object because the combined conditions ensure that `lz->result` is the only reference.

Reply:
@colesbury Thanks for the hint! I implementeded the suggestion

- I placed the logic in a single method `_Py_Reuse_Immutable_Object` so we can use the trick code for several other cases (e.g. `itertools.pairwise`, `enumerate`). The location and name of the new method can be changed though
- I re-ordered the three conditions because I suspect that `ob_ref_local` is most informative and fastest to check
- I am now checking `ob_ref_shared ` to be zero, but perhaps we should check on `Py_ARITHMETIC_RIGHT_SHIFT(Py_ssize_t, ob_ref_shared , _Py_REF_SHARED_SHIFT)`?